### ISSUE TITLE
Add multi-GPU support

### ISSUE THREAD

**Issue by philippmwirth** [lightly-ai/lightly#59]

**Comment by IgorSusmelj** [comment]
I think there are two ways to tackle the problem. Either we just run the loss computation on each node and then do backprop again on all of them. Or we use a more distributed method where each node collects negatives from the other nodes.

The first one should work out of the box using PyTorch lightning. We could just set the `accelerator` flag in PyTorch lightning to `ddp2` as described here: https://pytorch-lightning.readthedocs.io/en/latest/multi_gpu.html#distributed-data-parallel-2

The other approach would use a `gather` from torch to collect the negatives and then do the loss on each of the GPUs. Similar to here: https://github.com/Spijkervet/SimCLR/blob/master/modules/nt_xent.py

I would assume that for larger models and smaller batch sizes the two options are similar in terms of speed since the compute overhead of the forward pass of the model is significantly higher than the communication overhead. However, more efficient also for smaller models with larger batch sizes would be option 2. Approach 1 would use slightly more memory.

Since we have only one loss and the gather operation looks rather simple to implement we could try it out?

**Comment by philippmwirth** [comment]
I think we have to go for the second option, otherwise there is no benefit of having large batch sizes at all.

**Comment by IgorSusmelj** [comment]
I did run some tests and it looks like by setting `distributed_backend` in the pytorch lightning trainer to 'ddp' and `gpus=-1` already makes the code work well on a multi-GPU system. 
I'll add the information instead to the documentation to make it clear how to use lightly on multiple GPUs :)

**Comment by philippmwirth** [comment]
Does this hold for the `NTXEntLoss` as well? I fear that the computation of the loss is split over several GPUs and then the average (or sum) is taken for backpropagation. But this would mean that the number of negative samples is limited by the number of images which fit on a **single** GPU. Did you find anything about that?

**Comment by IgorSusmelj** [comment]
For the small benchmarks, I did it looks like our SimCLR model with 4 GPUs and each with batch size 32 converges very similar to our model trained with 1 GPU and batch size 128.

**Comment by IgorSusmelj** [comment]
There is also the option to use synchronized batch norm but it seems to not help but only slow down the training. I'll run some benchmarks in the background once I have time. From the experiments, 4 T4 GPUs are faster and cheaper than a single V100 GPU.

**Comment by IgorSusmelj** [comment]
I run an experiment using SimCLR.
Accuracy is 83% for 4 GPUs @ 128 batch size which is the same we got for batch size of 512

**Comment by IgorSusmelj** [comment]
I'll close this issue and create a new one for updating the documentation

**Comment by philippmwirth** [comment]
Distributed sinkhorn algorithm for `SwaV`, heavily inspired by https://github.com/facebookresearch/swav

```python
@torch.no_grad()
def sinkhorn(out: torch.Tensor, iterations: int = 3, epsilon: float = 0.05):
    """Distributed sinkhorn algorithm.

    As outlined in [0] and implemented in [1].
    
    [0]: SwaV, 2020, https://arxiv.org/abs/2006.09882
    [1]: https://github.com/facebookresearch/swav/ 

    Args:
        out:
            Similarity of the features and the SwaV prototypes.
        iterations:
            Number of sinkhorn iterations.
        epsilon:
            Temperature parameter.

    Returns:
        Soft codes Q assigning each feature to a prototype.
    
    """

    # get the world size in the distributed case
    world_size = 1
    if dist.is_initialized():
        world_size = dist.get_world_size()

    # get the exponential matrix and make it sum to 1
    Q = torch.exp(out / epsilon).t()
    sum_Q = torch.sum(Q)
    if world_size > 1:
        dist.all_reduce(sum_Q)
    Q /= sum_Q

    B = Q.shape[1] * world_size
    K = Q.shape[0] # number of prototypes

    for i in range(iterations):
        # normalize rows
        sum_of_rows = torch.sum(Q, dim=1, keepdim=True)
        if world_size > 1:
            dist.all_reduce(sum_of_rows)
        Q /= sum_of_rows
        Q /= K
        # normalize columns
        Q /= torch.sum(Q, dim=0, keepdim=True)
        Q /= B

    Q *= B
    return Q.t()
```

**Comment by philippmwirth** [comment]
This was added with #666.
